feat(terminal): show [Process exited] message in the statusline

cesar-guibo opened the pull request:
Removed the [Process exited] message that appeared in terminals with a finished process from the buffer contents and added it to the status line. However, to do that, it was necessary to expose struct terminal in src/nvim/terminal.h

fixes: #14986

BREAKING CHANGE: the Terminal type definition is not being supplied by "nvim/types_defs.h" anymore. It is being supplied by "nvim/terminal.h".

gpanders commented:
>Removed the [Process exited] message that appeared in terminals with a finished process from the buffer contents and added it to the status line.

What happens if a user has a custom `'statusline'` set?

>However, to do that, it was necessary to expose struct terminal in src/nvim/terminal.h

This shouldn't be necessary. The struct definition is "private" to the implementation. You can add new APIs if callers outside of the terminal implementation need to manipulate the internal data structure.

cesar-guibo commented:
> What happens if a user has a custom 'statusline' set?

In the cases that the user has a custom statusline set, it doesn't change anything. In those instances, the logic that is controlling the custom statusline isn't known.

> This shouldn't be necessary. The struct definition is "private" to the implementation. You can add new APIs if callers outside of the terminal implementation need to manipulate the internal data structure.

Okay, I will add an API instead of exposing the internal data structure

gpanders commented:
Before you spend more time on this, consider if this needs to be done in C. Do we already have the tools for users to put this information in the statusline themselves? Does `nvim_get_chan_info()` provide the relevant information? If so, users can just set `'stl'` to something that includes this without modifying the internal C code at all.

justinmk reviewed (changes requested):
> Does `nvim_get_chan_info()` provide the relevant information?

👍  as hinted in https://github.com/neovim/neovim/issues/14986#issuecomment-1676347598 , 

- The main missing piece is a way to get the result (error code) of the terminal process.
    - Add a `code` field to the result of `nvim_get_chan_info`.
    - ⚠️ Currently, when a process exits, Nvim destroys the job info, so `nvim_get_chan_info()` won't work.
        - To fix that, the _minimum_ interface is to change the internal channels/jobs list to keep ring buffer of the last N channels, instead of destroying the channel immediately.
        - Deprecate `nvim_list_chans`.
        - Deprecate `nvim_get_chan_info`.
        - Introduce `nvim_chan_get(channelId, opts): List<Dict>` 
- The default 'statusline' should show the exit code (e.g. `Exit code: 0`). Users with a custom statusline can get the exit code via `nvim_chan_get()`.

cesar-guibo commented:
> > Does `nvim_get_chan_info()` provide the relevant information?
> 
> 👍 as hinted in [#14986 (comment)](https://github.com/neovim/neovim/issues/14986#issuecomment-1676347598) ,
> 
>     * The main missing piece is a way to get the result (error code) of the terminal process.
>       
>       * Add a `code` field to the result of `nvim_get_chan_info`.
>       * ⚠️ Currently, when a process exits, Nvim destroys the job info, so `nvim_get_chan_info()` won't work.
>         
>         * To fix that, the _minimum_ interface is to change the internal channels/jobs list to keep ring buffer of the last N channels, instead of destroying the channel immediately.
>         * Deprecate `nvim_list_chans`.
>         * Deprecate `nvim_get_chan_info`.
>         * Introduce `nvim_chan_get(channelId, opts): List<Dict>`
> 
>     * The default 'statusline' should show the exit code (e.g. `Exit code: 0`). Users with a custom statusline can get the exit code via `nvim_chan_get()`.

Instead of altering the nvim API, wouldn't it be better to add the exit code as a statusline item (e.g. %e) for users with a custom statusline?

justinmk commented:
> Instead of altering the nvim API, wouldn't it be better to add the exit code as a statusline item (e.g. %e) for users with a custom statusline?

That seems like a hack, to avoid the work that is actually needed, which is that channel info should be available even after a channel/job exits. That is generally needed for various purposes. Whereas, adding some internal "pass around the exit code" logic is just yet another hack, and not an isolated one.

cesar-guibo commented:
Alright, I will get started on the API changes. 

A ```nvim_buf_get_termchan(handle): id``` will also be necessary since there currently aren't any ways to obtain the id of the terminal's channel.

zeertzjq commented:
`:h 'channel'`

justinmk commented:
> A `nvim_buf_get_termchan(handle): id` will also be necessary since there currently aren't any ways to obtain the id of the terminal's channel.

As @zeertzjq hinted, the buffer-local `&channel` option gives the terminal channel-id.

cesar-guibo commented:
Without any changes to the API it is possible to get the value of the exit code. The channel is not destroyed until the terminal is closed.

The current API passes the following test:
```
it('get exit status from terminal channel after channel is already closed #14986', function()
    command(([[terminal "%s" REP 5000 xxx]]):format(testprg('shell-test')))
    local exitstatus = -1
    while exitstatus < 0 do
      local handle = api.nvim_get_current_buf()
      local chan_id = api.nvim_get_option_value('channel', { buf = handle })
      local channel = api.nvim_get_chan_info(chan_id)
      assert(channel['id'] == chan_id)
      exitstatus = channel['exitstatus']
      sleep(10)
    end
    assert(exitstatus == 0)
  end)
```

cesar-guibo committed:
feat(terminal): show [Process exited] message in the statusline (#27195)

Removed the [Process exited] message that appeared in terminals with a
finished process from the buffer contents and added it to the statusline.

closes: #14986

cesar-guibo committed:
Merge branch 'master' into fix_14986
